### PR TITLE
docs: user actions conceptual guide

### PR DESCRIPTION
**Pull request by sixhobbits**
@mooreds - here's a more conceptual guide to user actions base on the blog post we did before.

For this one, we weren't completely sure if you wanted just a copy-paste from the blog to keep it in the docs as well, or something more conceptual so we might also need to do one or more of 

* revisit how we interlink the different user actions resources
* combine this new with the existing actioning users API docs
* create a new overview page that links to all of these and explains the goal of each so a reader can choose which to read

For context, we were going off [this github issue](https://github.com/FusionAuth/fusionauth-site/issues/25) and your slack message below for this.

image

### CODE REVIEW

**Comment by mark-robustelli** [comment]
@mooreds , I made some comments.  It would be very helpful to me if you took a look over this was well and made your comments so that I can see how we align and see what I missed if anything.

**Review by mark-robustelli — CHANGES_REQUESTED** [review]

File: site/docs/v1/tech/guides/user-actions.adoc [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+image::tutorials/user-actions/user-actions-edit-options.png[Defining an option and adding a localization to it, width=1200px, role=bottom-cropped]
+
+You can also add [uielement]#Localization# for the name of the User Action itself.
+
+image::tutorials/user-actions/user-actions-edit-localization.png[Adding a localization to the name of the User Action, width=1200px, role=bottom-cropped]
+
+You can refer to the link:/docs/v1/tech/apis/actioning-users[User Action event guide] for a full description of User Action event parameters.
+
+== Creating an API key
```
[the remark this answers is not in the document]
Comment (mooreds) [reply site/docs/v1/tech/guides/user-actions.adoc:51]: I love that idea.

File: site/docs/v1/tech/guides/user-actions.adoc [hunk 2]
```diff
@@ -0,0 +1,238 @@
+---
+layout: doc
+title: User Actions
+description: Using user actions
+navcategory: developer
+---
+:page-liquid:
+
```
Comment (mooreds) [inline site/docs/v1/tech/guides/user-actions.adoc:8]: Please add a link to this guide to the API pages:

https://fusionauth.io/docs/v1/tech/apis/actioning-users

and 

https://fusionauth.io/docs/v1/tech/apis/user-actions

File: site/docs/v1/tech/guides/user-actions.adoc [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+Navigate to [breadcrumb]#Tenants -> Your tenant#, and select the [breadcrumb]#Webhooks# tab. Make sure that the webhook is enabled. If you selected [uielement]#All tenants# on the webhook page, this checkbox will be disabled.
+
+image::tutorials/user-actions/user-actions-tenants-webhooks.png[Viewing the enabled webhooks on the Tenant page,width=1200px]
+
+Scroll down and make sure the [uielement]#user.action# event is enabled here too.
+
+image::tutorials/user-actions/user-actions-tenants-switch.png[Ensuring that the user.action Webhook event switch is enabled,width=1200px]
+
+== Executing the User Action
```
Comment (mooreds) [inline site/docs/v1/tech/guides/user-actions.adoc:143]:
```suggestion
== Actioning the User
```

File: site/docs/v1/tech/guides/user-actions.adoc [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    "userActionId": "6f4115c0-3db9-4734-aeda-b9c3f7dc4269"
+  }
+}
+----
+
+In this example, when the action is executed, the `actionee` will receive an email thanking them for their subscription.
+
+image::tutorials/user-actions/user-actions-email.png[Email confirmation from User Action event,width=1200px, role=bottom-cropped]
+
+== Querying Action Status on a User
```
Comment (mooreds) [inline site/docs/v1/tech/guides/user-actions.adoc:199]:
```suggestion
== Querying Applied User Actions
```

File: site/docs/v1/tech/guides/user-actions.adoc [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+     "userEmailingEnabled": true,
+     "sendEndEvent": true
+   }
+  }'
+----
+
+Here, `<YOUR_FUSIONAUTH_BASE_URL>` would be the URL of the FusionAuth instance and `<YOUR API KEY>` would be the API key noted earlier. Note that the email template Ids are included in the command. Read more about link:/docs/v1/tech/email-templates/configure-email[setting up email in FusionAuth here] and link:/docs/v1/tech/email-templates/email-templates[email templates in FusionAuth here].
+
+If you set the [field]#temporal# attribute to `true`, you can set an [field]#expiry# time for an action. This means that the action will automatically be removed from the user after the set expiry time. In this example, the [field]#sendEndEvent# field is set to `true` so that the user is notified via webhook when the access period has expired.
+
```
Comment (mooreds) [inline site/docs/v1/tech/guides/user-actions.adoc:87]:
```suggestion
You can also use any of our link:/docs/v1/tech/client-libraries/[client libraries] to make these API calls, but to keep this guide general, we'll use the API and curl.

```

**Review by mooreds — CHANGES_REQUESTED** [review]
I think this is an okay tutorial. I agree with all of Mark's comments. I also added some tactical suggestions, such as cropping images and rephrasing some sections.

But I think this can be so much better, and would like to make it so.

Here's the current table of contents:

* Types of User Actions
* Creating an API key
* Creating a User Action via the API
* Setting up a Webhook
* Executing the User Action
* Querying Action Status on a User
* Further reading

I'd like to update it to have the following sections/content:

# Types of User Actions

Add a table of the types:
Type | Use  | example
Time based | When you want to apply something for a given period of time. | subscription access, forum ban
Option based | When you want to apply certain value to a user, recording who did so, perhaps with comments | user surveyed and was happy/frustrated/indifferent 

And an explanation of how you use these in general:
* create a user action
* apply it to users via the API or admin UI (actioning)

Also mention/define any other jargon:

* action
* actioner
* actionee
* actioning a user
* temporal
... anything else

## Time Based User Actions

add a diagram of the flow:
 started -> modified (could loop to this) -> ended
                    \ (or)
                      ---> cancelled
### Time based action 
Example of the subscription from the blog.

## Option Based
Option
add a diagram of the flow: added -> removed

### Option based action 
Another example. How about an option that records interaction with a user and a customer service rep, assigns it a impact rating (high, medium, low) and includes a comment.

# What Happens After a User Action
## nothing, just a record that it happened and who did it.
## Webhooks
## Emails to the actioned user
# Setting Up User Actions
Tell a story here (or introduce it above and expand on it below). You already have the subscription example, so let's tell the story of Pied Piper expanding into media and building out both a subscription and a lightweight user happiness system for their customer service agents. Goal is to make it like this: https://fusionauth.io/docs/v1/tech/guides/multi-tenant in terms of telling a real world story.
## Creating an API key
## Creating a User Action via the API (create both types)

## Setting up a Webhook -- why? so we can feed the subscription and happiness info into our other systems? have an audit trail? 

## Set up the Email
Set up sending a thank you email to the user when their interaction has been recorded.
## Executing the User Action (execute both)
### Webhook Example
Show example of what webhook would look like when received and link to the webhook event documentation
### Show example of email
## Querying Action Status on a User (query both) and explain why you'd do this
# Localization -- extract all the localization stuff to here
- options
- name
- emails (point to email templates docs, no need to build this out entirely), but note that the email template is pulled based on the users preferred email
- anything else
# Further reading

**Comment by mooreds** [comment]
Please let me know when this is ready for a re-review.

File: site/docs/v1/tech/guides/user-actions.adoc [hunk 6]
```diff
@@ -0,0 +1,906 @@
...
+    Added-.->Removed
+{% endmermaid %}
+++++
+
+
+==== Survey Example
+
+Let's take an instantaneous Action example where a user gives feedback on their interaction with customer support by assigning a rating and giving a comment.
+
+Assume you have already created an instantaneous Action named "Feedback" in FusionAuth, with Options of "Bad", "Neutral", and "Good". Your user chooses "Good" in your feedback form and enters the comment "Problem solved quickly". When the form is saved, your code will call the Action API and create an Action instance for the User with the option "Good", and populate the [field]#comment# field. The [field]#actioner# of the instance will be set to the support User who helped the customer.
```
Comment (mooreds) [inline site/docs/v1/tech/guides/user-actions.adoc:180]: Wouldn't the support User be the actionee? And the customer be the actioner?

File: site/docs/v1/tech/guides/user-actions.adoc [hunk 7]
```diff
@@ -0,0 +1,906 @@
...
+
+Three separate APIs manage Actions. Each API has its own documentation.
+
+* link:/docs/v1/tech/apis/user-actions[Actions] — Defines an Action, updates it, and deletes it. The API path is `/api/user-action`.
+* link:/docs/v1/tech/apis/user-action-reasons[Action Reasons] — Defines the reason an Action is taken. The API path is `/api/user-action-reason`.
+* link:/docs/v1/tech/apis/actioning-users[Action instances] — Applies an existing Action to a User, optionally with a Reason. Can also update or cancel the Action instance. The API path is `/api/user/action`.
+
+Actions and Action Reasons can be managed on the FusionAuth admin UI. Only Action instances require you to use the API: You cannot apply an Action to a User on the website.
+
+It is faster to use FusionAuth API wrappers rather than make HTTP calls directly. You can read how to use API wrappers in the link:/docs/v1/tech/client-libraries/[client library guide] before continuing. This guide uses the TypeScript client library.
```
Comment (mooreds) [inline site/docs/v1/tech/guides/user-actions.adoc:202]: Please call them 'client libraries' not API wrappers.

File: site/docs/v1/tech/guides/user-actions.adoc [hunk 8]
```diff
@@ -0,0 +1,906 @@
...
+** [field]#Application#: `PiedPiper`
+** [field]#Roles#: `customer`
+
+image::guides/user-actions/reader-user.png[Create a Subscriber User in FusionAuth,width=1200,role=bottom-cropped]
+
+Record the [field]#User Id# of both the Users you created to use later.
+
+=== Create an API Key
+
+You now have an Application with two Users.
```
Comment (mooreds) [inline site/docs/v1/tech/guides/user-actions.adoc:395]:
```suggestion
You now have an Application with two registered Users.
```

File: site/docs/v1/tech/guides/user-actions.adoc [hunk 9]
```diff
@@ -0,0 +1,906 @@
...
+
+To apply Actions using the API, we need to create an API Key. In reality, you should grant as few privileges as possible to an API Key (principle of least privilege), but you'll make a skeleton key in this tutorial to save time.
```
Comment (mooreds) [inline site/docs/v1/tech/guides/user-actions.adoc:397]:
```suggestion
To apply Actions using the API, you need to create an API Key. In reality, you should grant as few privileges as possible to an API Key (principle of least privilege), but you'll make a key with all privileges in this tutorial to save time.
```

File: site/docs/v1/tech/guides/user-actions.adoc [hunk 10]
```diff
@@ -0,0 +1,906 @@
...
+
+Since your Actions will rely on calling webhooks, you're going to create the webhooks first. Your first webhook will notify Intercom that a new user has subscribed and should be sent the onboarding series of emails that explain how to use all the paid features of PiedPiper. All our webhooks in this tutorial are sent to fake localhost versions of these real companies.
+
+* Navigate to [breadcrumb]#Settings -> Webhooks# and add icon:plus[role=ui-button green,type=fas] a webhook.
+** [field]#Id#: `55934340-3c92-410a-b361-40fb324ed412`
+** [field]#URL#: `\http://host.docker.internal:3000/intercom`
+** Scroll down and ensure that the [field]#user.action# event is enabled.
+* Save icon:save[role=ui-button blue,type=fas] the webhook.
+
+image::guides/user-actions/create-webhook.png[Create a Webhook in FusionAuth,width=1200,role=bottom-cropped]
```
Comment (mooreds) [inline site/docs/v1/tech/guides/user-actions.adoc:480]: Why is the `Event` tab cut off? Seems like you'd want to show it.

File: site/docs/v1/tech/guides/user-actions.adoc [hunk 11]
```diff
@@ -0,0 +1,906 @@
...
+Enabling the webhooks in two places gives you fine-grained control across Tenants. More information on webhooks is available link:/docs/v1/tech/events-webhooks/#overview[here].
+====
+
+==== Create Subscription Action
+
+Now you can create the subscription and banning Actions to apply to the user in our PiedPiper code. They're both temporal Actions.
+
+[NOTE]
+====
+You'll continue using the FusionAuth admin UI to create objects in this tutorial. If you think it would be faster in future to create Actions in code, see this previous https://fusionauth.io/blog/2023/04/20/using-user-actions#creating-the-user-action[guide] that demonstrates creating Actions in the terminal.
```
Comment (mooreds) [inline site/docs/v1/tech/guides/user-actions.adoc:510]:
```suggestion
You'll continue using the FusionAuth admin UI to create objects in this tutorial. You can also use client libraries and the API. Here's a link:/blog/2023/04/20/using-user-actions#creating-the-user-action[blog post which illustrates this functionality].
```

File: site/docs/v1/tech/guides/user-actions.adoc [hunk 12]
```diff
@@ -0,0 +1,906 @@
...
+
+=== Create a Subscriber User (Actionee)
+
+* Under [breadcrumb]#Users#, click the icon:plus[role=ui-button green,type=fas] button to add a User.
+* Enter the values:
+** [field]#Email#: `reader@example.com`
+** Disable [field]#Send email to set up password# to manually set the password.
+*** [field]#Password#: `password`
+*** [field]#Confirm#: `password`
+** [field]#Languages#: `Esperanto` (Note that you have to enter the text, wait for a popup to appear, then click it to confirm the entry.)
```
Comment (mooreds) [inline site/docs/v1/tech/guides/user-actions.adoc:383]: :)

File: site/docs/v1/tech/guides/user-actions.adoc [hunk 13]
```diff
@@ -0,0 +1,906 @@
...
+Let's start testing by applying the subscription Action to the user. In reality, your app would do this in code once the user has paid, but for now we'll do it in a new terminal.
+
+[NOTE]
+====
+You'll need to install `curl` if it's not already installed.
+====
+
+In the following code, you will replace the values of [filed]#actioneeUserId# and [field]#actionerUserId# with the values you recorded earlier for the reader and administrator users respectively.
+
+To test out the workflow, you can let the subscription expire after 60 seconds. From the https://fusionauth.io/dev-tools/date-time[FusionAuth Date-Time tool], copy the [field]#Milliseconds# value, add `60000` (60 seconds) to it, and paste it into the `"expiry"` field below. This will ensure the subscription action expires immediately. If you're on Linux, use the Option 2 code snippet to set the [field]#expiry# value automatically.
```
Comment (mooreds) [inline site/docs/v1/tech/guides/user-actions.adoc:716]:
```suggestion
To test out the workflow, you can let the subscription expire after 60 seconds. From the https://fusionauth.io/dev-tools/date-time[FusionAuth Date-Time tool], copy the [field]#Milliseconds# value, add `60000` (60 seconds) to it, and paste it into the `"expiry"` field below. This will ensure the subscription action expires quickly. If you're on Linux, use the Option 2 code snippet to set the [field]#expiry# value automatically.
```

File: site/docs/v1/tech/guides/user-actions.adoc [hunk 14]
```diff
@@ -0,0 +1,906 @@
...
+      "comment": "Paid for the news",
+      "emailUser": true,
+      "expiry": '"$(($(date +%s) * 1000 + 60000))"',
+      "userActionId": "38bf18dd-6cbc-453d-a438-ddafe0daa1b0",
+      "reasonId": "ae080fe4-5650-484f-807b-c692e218353d"
+    }
+  }'
+----
+
+You should receive a 200 status code and a response that looks like the following.
```
Comment (mooreds) [inline site/docs/v1/tech/guides/user-actions.adoc:760]:
```suggestion
You should receive a `200` status code and a response that looks like the following.
```

**Comment by mooreds** [comment]
Also, please pull any code in from a GH repo, using an include. I know you built on the 5 minute setup guide, but see there were a few additions.

I'm happy to create a GH repo under the FusionAuth org, Please set up a ritza repo and I can pull the code over.

**Comment by mooreds** [comment]
Is this ready for re-review, @sixhobbits ?

**Comment by sixhobbits** [comment]
@mooreds - ready for you again. The only thing we're not sure about re the include is that previously we explained some bits of it in sections, but we can't pull in sections using an include, so this ends up with a 'wall of code' that is a bit hard to follow 

at the moment we are still including some sections manually to explain the mock intercom, slack, and expire sections. Let me know if you prefer us to include the entire file (https://github.com/ritza-co/fusionauth-example-user-actions-guide/blob/main/app.js) and then try explain it all at once rather.

here's the repo that needs to moved over: 
https://github.com/ritza-co/fusionauth-example-user-actions-guide

**Comment by mooreds** [comment]
@sixhobbits you can actually pull in particular sections of code from a remote repo. Let's use that approach. 

Here's an example: https://raw.githubusercontent.com/FusionAuth/fusionauth-site/master/site/docs/v1/tech/getting-started/_5-minute-store-user-object.adoc

Let me know when that is done and I'll move the repository over.

**Comment by mooreds** [comment]
Is this ready for re-review?

**Comment by sixhobbits** [comment]
@mooreds yes, all yours! I'm sometimes the bottleneck on pressing the 'request review' button as @rideam doesn't have that option - not sure if you know how to grant him access to that?

**Comment by mooreds** [comment]
I get it!

@rideam , I just gave you the triage role (I think you have to accept it). That should allow you to push the 'request review' button. Please do so whenever a piece is ready for review.

**Review by mooreds — APPROVED** [review]
LGTM

**Comment by mooreds** [comment]
Closing under https://github.com/FusionAuth/fusionauth-site/pull/2510
